mllib/src/main/scala/org/apache/spark/mllib/linalg/Vectors.scala

```diff
@@ -77,6 +77,11 @@ sealed trait Vector extends Serializable {
         // refer to {@link java.util.Arrays.equals} for hash algorithm
         val bits = java.lang.Double.doubleToLongBits(value)
         result = 31 * result + (bits ^ (bits >>> 32)).toInt
+        i += 1
+        // only scan the first 16 nonzeros
+        if (i > 16) {
+          return result
```

Comment:
note that return in a closure is pretty expensive because it triggers an exception. this might be much slower than before. you probably want to pass the max number of things to run into foreachActive.